src/garage/experiment/local_tf_runner.py

```diff
@@ -319,6 +323,39 @@ def train(self,
             pause_for_plot=pause_for_plot,
             start_epoch=0)
 
+        return self._train()
+
+    def _step_epochs(self):
+        """Magic generator for training.
+
+        This function serves as a generator. It is used to separate
+        services such as snapshotting, sampler control from the actual
+        training loop.
+
+        """
+        self.start_worker()
+
+        self.start_time = time.time()
+        self.step_epochs_args = SimpleNamespace(
+            itr=self.train_args.start_epoch * self.train_args.n_epoch_cycles,
+            path=None)
+
+        for epoch in range(self.train_args.start_epoch,
+                           self.train_args.n_epochs):
+            self.itr_start_time = time.time()
+
+            with logger.prefix('epoch #%d | ' % epoch):
+                yield
+
+                self.save(
+                    epoch, self.step_epochs_args.path
+                    if self.train_args.store_paths else None)
+                self.log_diagnostics(self.train_args.pause_for_plot)
+                logger.dump_all(self.step_epochs_args.itr)
+                tabular.clear()
+
+        self.shutdown_worker()
```

Comment:
It would be nice for `shutdown_worker()` to be called in a finally clause, so that the workers will get shutdown even if the loop using this generator throws an exception. This can be in a later PR though.

Please either make this change in this PR, or create a github issue.